### PR TITLE
Add k8s test

### LINKED ISSUE #205: Add tests for kubernetes side of the CNI code.

### PR DESCRIPTION
**Pull request by tomdee**
Fixes #205

### CODE REVIEW

**Comment by tomdee** [comment]
First commit is being reviewed under #228

**Comment by tomdee** [comment]
I'm assuming you're still looking at this? @robbrockbank

File: test_utils/utils.go [hunk 1]
```diff
@@ -65,7 +138,7 @@ func RunIPAMPlugin(netconf, command, args string) (types.Result, int) {
 	return result, exitCode
 }
 
-func CreateContainer(netconf string) (container_id, netnspath string, session *gexec.Session, contVeth netlink.Link, contAddr []netlink.Addr, contRoutes []netlink.Route, err error) {
+func CreateContainer(netconf string, name string) (container_id, netnspath string, session *gexec.Session, contVeth netlink.Link, contAddr []netlink.Addr, contRoutes []netlink.Route, err error) {
```
Comment (robbrockbank) [inline test_utils/utils.go]: should name be k8sName or something? Not immediately obvious it's a k8s only thing

File: test_utils/utils.go [hunk 2]
```diff
@@ -16,13 +21,81 @@ import (
...
+var kapi etcdclient.KeysAPI
+
+func init() {
+
+	cfg := etcdclient.Config{Endpoints: []string{"http://127.0.0.1:2379"}}
+	c, _ := etcdclient.New(cfg)
+	kapi = etcdclient.NewKeysAPI(c)
+}
+
+// Delete everything under /calico from etcd
```
Comment (robbrockbank) [inline test_utils/utils.go]: All nice function to have, but again - would be useful to have in our testutils :-)

File: test_utils/utils.go [hunk 3]
```diff
@@ -16,13 +21,81 @@ import (
...
+func WipeEtcd() {
+	_, err := kapi.Delete(context.Background(), "/calico", &etcdclient.DeleteOptions{Dir: true, Recursive: true})
+	if err != nil && !etcdclient.IsKeyNotFound(err) {
+		panic(err)
+
+	}
+
+}
+
+func WipeK8sPods() {
```
Comment (robbrockbank) [inline test_utils/utils.go]: Maybe a comment here.  Looks like everything runs in a test namespace - is that true?

File: test_utils/utils.go [hunk 4]
```diff
@@ -16,13 +21,81 @@ import (
...
+	for _, pod := range pods.Items {
+		err = clientset.Pods("test").Delete(pod.Name, &v1.DeleteOptions{})
+
+		if err != nil {
+			panic(err)
+		}
+	}
+}
+
+func PreCreatePool(pool string) {
```
Comment (robbrockbank) [inline test_utils/utils.go]: Something similar in testutils.client    in libcalico-go

File: calico_cni_test.go [hunk 5]
```diff
@@ -27,13 +29,25 @@ import (
 // vary the MTU
 // Existing endpoint
 
+var calicoClient *client.Client
+
+func init() {
```
Comment (robbrockbank) [inline calico_cni_test.go]: Something similar in testutils.client    in libcalico-go

File: calico_cni_k8s_test.go [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+					panic(err)
+				}
+				clientset, err := kubernetes.NewForConfig(config)
+
+				if err != nil {
+					panic(err)
+				}
+				name := fmt.Sprintf("run%d", rand.Uint32())
+				interfaceName := k8s.VethNameForWorkload(fmt.Sprintf("test.%s", name))
+				_, err = clientset.Pods("test").Create(&v1.Pod{
```
Comment (robbrockbank) [inline calico_cni_k8s_test.go]: nit:  Given we rely on the delete code deleting the entries in this namespace perhaps nicer to have this namespace as a const.

**Comment by robbrockbank** [comment]
@tomdee:  It LGTM, but would be nice to resolve some of the niggles that i've mentioned.  Similar to the previous review - there's some useful function that is either already in the testutils or would be better in the testutils.

File: test_utils/utils.go (hunk 1 above)
Comment (tomdee) [inline test_utils/utils.go]: Done

File: test_utils/utils.go (hunk 3 above)
Comment (tomdee) [inline test_utils/utils.go]: Done

File: test_utils/utils.go [hunk 7]
```diff
@@ -135,7 +162,12 @@ func CreateContainer(netconf string) (container_id, netnspath string, session *g
 
 	// Set up the env for running the CNI plugin
 	//TODO pass in the env properly
-	cni_env := fmt.Sprintf("CNI_COMMAND=ADD CNI_CONTAINERID=%s CNI_NETNS=%s CNI_IFNAME=eth0 CNI_PATH=dist", container_id, netnspath)
+	var k8s_env = ""
+	if name != "" {
+
```
Comment (tomdee) [inline test_utils/utils.go]: Done

File: test_utils/utils.go (hunk 2 above)
Comment (tomdee) [inline test_utils/utils.go]: Yeah, but that uses etcdctl, bleurghh

File: test_utils/utils.go (hunk 4 above)
Comment (tomdee) [inline test_utils/utils.go]: That one's a little more "feature rich". I'm happy with a little duplication for the moment to keep the code simpler.

File: calico_cni_test.go (hunk 5 above)
Comment (tomdee) [inline calico_cni_test.go]: Cool, using it.

**Comment by tomdee** [comment]
Fixed up all comments (except for using the etcdctl based WipeData from libcalico-go - I don't have time to fix that up "upstream"), so merging.

**Pull request merged**